tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -13236,6 +13205,164 @@ void ExecutionTest::WaveSizeTest() {
+                        "}\r\n";
+
+  struct WaveSizeTestData {
+    uint32_t count;
+  };
+
+  for (UINT minShaderWaveSize = 4; minShaderWaveSize <= maxWaveSize;
+       minShaderWaveSize *= 2) {
+    for (UINT maxShaderWaveSize = minShaderWaveSize * 2;
+         maxShaderWaveSize <= 128; minShaderWaveSize *= 2) {
+      for (UINT prefShaderWaveSize = minShaderWaveSize;
+           prefShaderWaveSize <= maxShaderWaveSize; prefShaderWaveSize *= 2) {
+        // format compiler args
+        char compilerOptions[64];
+        VERIFY_IS_TRUE(
+            sprintf_s(compilerOptions, sizeof(compilerOptions),
+                      "-D MINWAVESIZE=%d -D MAXWAVESIZE=%d -D PREFWAVESIZE=%d",
+                      minShaderWaveSize, maxShaderWaveSize,
+                      prefShaderWaveSize) != -1);
+
+        // run the shader
+        std::shared_ptr<ShaderOpTestResult> test = RunShaderOpTestAfterParse(
+            pDevice, m_support, "WaveSizeTest",
+            [&](LPCSTR Name, std::vector<BYTE> &Data, st::ShaderOp *pShaderOp) {
+              VERIFY_IS_TRUE((0 == strncmp(Name, "UAVBuffer0", 10)));
+              pShaderOp->Shaders.at(0).Arguments = compilerOptions;
+              pShaderOp->Shaders.at(0).Text = waveSizeTestShader;
+
+              VERIFY_IS_TRUE(sizeof(WaveSizeTestData) * MAX_WAVESIZE <=
+                             Data.size());
+              WaveSizeTestData *pInData = (WaveSizeTestData *)Data.data();
+              memset(pInData, 0, sizeof(WaveSizeTestData) * MAX_WAVESIZE);
+            },
+            ShaderOpSet);
+
+        // verify expected values
+        MappedData dataUav;
+        WaveSizeTestData *pOutData;
+
+        LogCommentFmt(
+            L"Verifying test result for these shader wave size ranges:");
+        LogCommentFmt(L"Minimum shader wave size: %d", minShaderWaveSize);
+        LogCommentFmt(L"Maximum shader wave size: %d", maxShaderWaveSize);
+        LogCommentFmt(L"Preferred shader wave size: %d", prefShaderWaveSize);
+
+        bool AcceptedByRuntime = TestShaderRangeAgainstRequirements(
+            minShaderWaveSize, maxShaderWaveSize, minWaveSize, maxWaveSize);
+        if (AcceptedByRuntime) {
+
+          test->Test->GetReadBackData("UAVBuffer0", &dataUav);
+          VERIFY_ARE_EQUAL(sizeof(WaveSizeTestData) * MAX_WAVESIZE,
+                           dataUav.size());
+          pOutData = (WaveSizeTestData *)dataUav.data();
+
+          for (unsigned i = 0; i < MAX_WAVESIZE; i++) {
+            if (!VERIFY_IS_TRUE(pOutData[i].count >= minWaveSize &&
+                                pOutData[i].count <= maxWaveSize))
+              break;
+            if (prefShaderWaveSize >= minWaveSize &&
+                prefShaderWaveSize <= maxWaveSize) {
+
+              if (!VERIFY_ARE_EQUAL(pOutData[i].count, prefShaderWaveSize))
+                break;
+            }
+          }
+        } else {
+          VERIFY_ARE_EQUAL(0u, dataUav.size());
```

Comment:
I would add a logging comment here that the wave size combination was not supported.

Also, I am not sure if the test will be rejected silently to even get here. I believe if the shader is not accepted it will fail the test and stop executing any further, but I might be wrong. If that happens, we might need to adjust the test to run just on the supported wave size range.

Reply:
Done!